### DISCUSSION TITLE
ViewConstruct in tm1py

### DISCUSSION THREAD

**Discussion by wimgielis** [cubewise-code/tm1py#745] [General]
**Describe what did you try to do with TM1py**
I am trying to run a ViewConstruct TI-equivalent in REST API. It should be done for all public cube views over all cubes that start with a certain name pattern.

**Describe what's not working the way you expect**
Does TM1py have a way to do this ? I did not find it and moreover, I did not find references to ViewConstruct or Construct within the $metadata document.

Latest versions.

I have this working code already, only ViewConstruct is missing:

``` python
from TM1py.Services import TM1Service

def viewconstruct_all_matching_public_views():

    ####################################################
    # Author: Wim Gielis
    # Date: 07-03-2021
    # Purpose:
    # - execute a ViewConstruct for every public cube view that starts with "ViewConstruct_*"
    # - 
    ####################################################

    # TM1 connection settings (IntegratedSecurityMode = 1 for now)
    ADDRESS = 'localhost'
    USER = 'wim'
    PWD = ''
    PORT = 8001
    SSL = False

    RESULT_FILE = r'D:\viewconstruct.txt'

    # execute a ViewConstruct only for view names that match the following pattern
    mask = "ViewConstruct_"

    log_lines = []
    public_views = {}

    tm1 = TM1Service(address=ADDRESS, port=PORT, user=USER, password=PWD, namespace='', gateway='', ssl=SSL)

    # Iterate through cubes to retrieve certain public views
    cube_names = tm1.cubes.get_all_names()
    for cube_name in cube_names:
        private_view_names, public_views_names = tm1.cubes.views.get_all_names(cube_name=cube_name)
        if len(public_views_names):
            matching_public_views = list(filter(lambda vw: vw.startswith(mask), public_views_names))
            for k in matching_public_views:
                log_lines.append("{}\t{}".format(cube_name, k))

    with open(RESULT_FILE, 'w', encoding='utf-8') as file:
        file.write("\n".join(log_lines))
        file.close()

if __name__ == "__main__":
    viewconstruct_all_matching_public_views()
```

**Comment by wimgielis** [comment]
The equivalent TI code:

```
`vPattern = 'viewconstruct_*';
c0 = 'temp';

c = 1;
While( c <= Dimsiz( '}Cubes' ));
   vCube = Dimnm( '}Cubes', c );

   vDim_CubeViews = '}Views_' | vCube;
   If( DimensionExists( vDim_CubeViews ) > 0 );

      If( Dimsiz( vDim_CubeViews ) > 0 );

         SubsetCreateByMDX( c0, '{TM1FilterByPattern( TM1SubsetAll( [' | vDim_CubeViews | '] ), "' | vPattern | '")}', vDim_CubeViews, 1 );
         SubsetElementInsert( vDim_CubeViews, c0, Dfrst( vDim_CubeViews ), 1 );
         SubsetElementDelete( vDim_CubeViews, c0, 1 );

         x = 1;
         While( x <= SubsetGetSize( vDim_CubeViews, c0 ));

            vViewName = SubsetGetElementName( vDim_CubeViews, c0, x );
            ViewConstruct( vCube, vViewName );

            x = x + 1;
         End;

      EndIf;

   EndIf;

   c = c + 1;
End;
```

**Comment by MariusWirtz** [comment]
Hi @wimgielis,

some TI functions have no equivalent in the REST API. `ViewConstruct` is one of them.

If your code runs as admin, you can just execute an unbound TI process with a `viewconstruct `statement in it.

``` python
from TM1py import Process
from TM1py.Services.TM1Service import TM1Service

with TM1Service(address="", port=12354, ssl=True, user="admin", password="apple") as tm1:
    process = Process(name="", prolog_procedure="ViewConstruct('c1', 'Default');")
    tm1.processes.execute_process_with_return(process=process)
```

Just out of curiosity, if you execute the view wouldn't that trigger caching as well?

``` python
from TM1py.Services.TM1Service import TM1Service

with TM1Service(address="", port=12354, ssl=True, user="admin", password="apple") as tm1:
    tm1.cells.create_cellset_from_view(cube_name='c1', view_name='Default', private=False)

```

If this is a different kind of caching, perhaps we could add a `view_construct` function to TM1py (perhaps to the `ViewService`) that calls the unbound TI process behind the scenes.

**Comment by TM1Fanboy** [comment]
I have not really often used ViewConstruct but unless it ignores VMM and VMT and stores it in memory anyway, it should IMHO (but untested from my side) produce the same result. 
[IBM documentation](https://www.ibm.com/support/knowledgecenter/en/SSD29G_2.0.0/com.ibm.swg.ba.cognos.tm1_ref.2.0.0.doc/r_tm1_ref_tifun_viewconstruct.html) is (as always) not helpful. 

![ViewConstructDocumentation](https://user-images.githubusercontent.com/8442014/110308507-7c60b500-8000-11eb-810b-4fe0d7f44155.PNG)

**Comment by wimgielis** [comment]
Thanks both. I’ll play around with it and get back to you.



Op ma 8 mrt. 2021 om 11:22 schreef Christoph Hein <notifications@github.com>

> I have not really often used ViewConstruct but unless it ignores VMM and
> VMT and stores it in memory anyway, it should IMHO (but untested from my
> side) produce the same result.
> IBM documentation
> <https://www.ibm.com/support/knowledgecenter/en/SSD29G_2.0.0/com.ibm.swg.ba.cognos.tm1_ref.2.0.0.doc/r_tm1_ref_tifun_viewconstruct.html>
> is (as always) not helpful.
>
> [image: ViewConstructDocumentation]
> <https://user-images.githubusercontent.com/8442014/110308507-7c60b500-8000-11eb-810b-4fe0d7f44155.PNG>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cubewise-code/tm1py/issues/504#issuecomment-792652311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDHULIF24PEUS3HYWHT2UTTCSQOZANCNFSM4YXWXEPA>
> .
>
-- 

------

Best regards / Beste groeten,

Wim Gielis
MS Excel MVP 2011-2014
https://www.wimgielis.com <http://www.wimgielis.be>

**Comment by wimgielis** [comment]
Hello,

I did some testing on an simple TM1 model with 1 cube containing lots of
dimensions. There's no data but just a rule: [] = N: 5; and no skipcheck
and no feeders.
When I organize the dimensions in rows and columns it takes about 5-10
seconds to recalculate in Architect / Perspectives, Cube viewer.
VMM = 2000000 (2GB), VMT = 2 (sec). The Performance Monitor is on.
Still I don't see the Number of Stored Views increase in the }StatsByCube
cube. It remains 0. The view recalc takes 5-10 sec. and there's plenty of
memory in the VMM so I don't understand it.
A TI process with ViewConstruct to the same view in the Prolog tab is very
fast (instantaneous), probably because the view was already opened to be
able to save it as a public view. I did not create the view in TI for this
simple test.

Am I missing something here ? I also looked at, to no avail:
https://blog.octanesolutions.com.au/performance-enhancements-using-tm1-cached-views

------

Best regards / Beste groeten,

Wim Gielis
MS Excel MVP 2011-2014
https://www.wimgielis.com <http://www.wimgielis.be>


Op ma 8 mrt. 2021 om 13:50 schreef Wim Gielis ***@***.***>:

**Comment by TM1Fanboy** [comment]
Hey Wim, maybe I am missing something but a view without data entries or feeders cannot get stored in memory I would think. No data to store and feeder flags to store. Everything else is recalculated every time the view is opened. 

Have you tried entering some data by spreading from root and see if that makes a difference? Or Enter something into a single cell and feed everything from there.

**Comment by wimgielis** [comment]
Good point, haven’t tried that.


Op wo 10 mrt. 2021 om 20:28 schreef Christoph Hein ***@***.***
>

> Hey Wim, maybe I am missing something but a view without data entries or
> feeders cannot get stored in memory I would think. No data to store and
> feeder flags to store. Everything else is recalculated every time the view
> is opened.
>
> Have you tried entering some data by spreading from root and see if that
> makes a difference? Or Enter something into a single cell and feed
> everything from there.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cubewise-code/tm1py/issues/504#issuecomment-795956474>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEDHULNJYQIJHUWYHHBQCULTC7B5XANCNFSM4YXWXEPA>
> .
>
-- 

------

Best regards / Beste groeten,

Wim Gielis
MS Excel MVP 2011-2014
https://www.wimgielis.com <http://www.wimgielis.be>

**Comment by wimgielis** [comment]
Hi guys,

I am unable to force that a view is stored / cached by TM1 (measuring by entries in the }StatsByCube cube). I played around with VMM/VMT, not with Python but just in TM1 Architect.
If anyone wants to have a look, I'll add my TM1 model. It's only 1 cube, 1 TI, only its dimensions, nothing more. Very very small, nothing that should not be there.

Thanks for any pointers !
[tm1.zip](https://github.com/cubewise-code/tm1py/files/6130616/tm1.zip)

**Comment by wimgielis** [comment]
Username to log in is wim without a password ;-)

Op vr 12 mrt. 2021 om 15:29 schreef Wim Gielis ***@***.***>:
